### PR TITLE
Add snowplow-cli data products

### PR DESCRIPTION
**Pull request by gleb-lobov**

### CODE REVIEW

File: docs/pipeline-components-and-applications/cli/index.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
+---
+title: Snowplow CLI
+date: 2024-11-25
+sidebar_label: Snowplow CLI
+sidebar_position: 999
+---
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+`snowplow-cli` means to bring Snowplow Console into the command line. It allows you to download your data structures and data products to yaml/json files and publish them back to console. This enables the git-ops workflow, with reviews and brancing.
```
Comment (cksnp) [inline docs/pipeline-components-and-applications/cli/index.md:10]:
```suggestion
`snowplow-cli` brings data management elements of Snowplow Console into the command line. It allows you to download your data structures and data products to yaml/json files and publish them back to console. This enables git-ops-like workflows, with reviews and brancing.
```

File: docs/pipeline-components-and-applications/cli/index.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+# Install
+
+Snowplow CLI can be installed with [homebrew](https://brew.sh/)
```
Comment (cksnp) [inline docs/pipeline-components-and-applications/cli/index.md:14]:
```suggestion
Snowplow CLI can be installed with [homebrew](https://brew.sh/):
```

File: docs/pipeline-components-and-applications/cli/index.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+Verify the installation with
+```
+./snowplow-cli --help
+```
+
+# Configure
+
+You will need three values.
+
+API Key Id and API Key Secret are generated from the [credentials section](https://console.snowplowanalytics.com/credentials) in BDP Console.
```
Comment (cksnp) [inline docs/pipeline-components-and-applications/cli/index.md:42]:
```suggestion
An API Key Id and the corresponding API Key (secret), which are generated from the [credentials section](https://console.snowplowanalytics.com/credentials) in BDP Console.
```

File: docs/pipeline-components-and-applications/cli/index.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+Organization Id can be retrieved from the URL immediately following the .com when visiting BDP console:
```
Comment (cksnp) [inline docs/pipeline-components-and-applications/cli/index.md:44]:
```suggestion
The organization ID, which can be retrieved from the URL immediately following the .com when visiting BDP console:
```

File: docs/pipeline-components-and-applications/cli/index.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  </TabItem>
+</Tabs>
+
+Snowplow CLI defaults to yaml format. It can be changed to json by either providing a `--output-format json` flag or setting the `output-format: json` config value. It will work for all commands where it matters, not only for `generate`.
+
+
+# Use cases
+
+- [Manage your data structures with snowplow-cli](/docs/understanding-tracking-design/managing-your-data-structures/cli/index.md)
+- [Set up a github CI/CD pipeline to manage data structures](/docs/recipes/recipe-data-structures-in-git/index.md)
```
Comment (cksnp) [inline docs/pipeline-components-and-applications/cli/index.md:85]:
```suggestion
- [Set up a github CI/CD pipeline to manage data structures and data products](/docs/recipes/recipe-data-structures-in-git/index.md)
```

File: docs/recipes/recipe-data-structures-in-git/index.md [hunk 6]
```diff
@@ -16,19 +16,19 @@ The Snowplow Console's UI offers excellent facilities to get started quickly wit
 
 A common solution when faced with these requirements is to move management to some form of version control platform (github/gitlab). This opens up an entire ecosystem of tools and patterns enabling all manner of custom workflows.
 
-We have built [snowplow-cli](/docs/understanding-tracking-design/managing-your-data-structures/cli/) to help you bridge the gap between these repository based workflows and BDP Console.
+We have built [Snowplow CLI](/docs/pipeline-components-and-applications/cli/index.md) to help you bridge the gap between these repository based workflows and BDP Console.
```
Comment (cksnp) [inline docs/recipes/recipe-data-structures-in-git/index.md:19]:
```suggestion
We have built [Snowplow CLI](/docs/pipeline-components-and-applications/cli/index.md) to help you bridge the gap between these repository-based workflows and BDP Console.
```

File: docs/recipes/recipe-data-structures-in-git/index.md [hunk 7]
```diff
@@ -360,7 +360,206 @@ Validation has passed. Now our colleagues can feedback on our changes and if eve
...
+
+### Create a source applications
+
+First, we'll create a source application to represent our website that will send the `login` event we defined earlier.
+
+```bash
+snowplow-cli dp generate --source-app website
+```
+:::note
+`dp` is an alias for `data-products`. Source apps and event specsifications are also managed by this command
```
Comment (cksnp) [inline docs/recipes/recipe-data-structures-in-git/index.md:375]:
```suggestion
`dp` is an alias for `data-products`. Source applications and event specifications are also managed by this command
```

File: docs/recipes/recipe-data-structures-in-git/index.md [hunk 8]
```diff
@@ -360,7 +360,206 @@ Validation has passed. Now our colleagues can feedback on our changes and if eve
...
+:::
+
+This should provide the following output
+```
+INFO generate wrote kind="source app" file=data-products/source-apps/website.yaml
+```
+
+The generated file is written to our default `data-products/source-apps`. Help for all the arguments available to `generate` is available by running `snowplow-cli dp generate --help`.
```
Comment (cksnp) [inline docs/recipes/recipe-data-structures-in-git/index.md:383]:
```suggestion
The generated file is written to the default `data-products/source-apps` directory. Help for all the arguments available to `generate` is available by running `snowplow-cli dp generate --help`.
```

File: docs/recipes/recipe-data-structures-in-git/index.md [hunk 9]
```diff
@@ -360,7 +360,206 @@ Validation has passed. Now our colleagues can feedback on our changes and if eve
...
+
+When we're happy with the proposed changes, we can publish by removing the `--dry-run` flag:
+
+```bash
+snowplow-cli dp publish
+```
+
+After publishing, you'll be able to see your new source application in the BDP Console UI.
+
+### Create a data product and an event spec
```
Comment (cksnp) [inline docs/recipes/recipe-data-structures-in-git/index.md:435]:
```suggestion
### Create a data product and an event specification
```

File: docs/recipes/recipe-data-structures-in-git/index.md [hunk 10]
```diff
@@ -360,7 +360,206 @@ Validation has passed. Now our colleagues can feedback on our changes and if eve
...
+
+Let's now create a data product and an event spec by running the following command
```
Comment (cksnp) [inline docs/recipes/recipe-data-structures-in-git/index.md:437]:
```suggestion
Let's now create a data product and an event specification by running the following command
```

File: docs/recipes/recipe-data-structures-in-git/index.md [hunk 11]
```diff
@@ -360,7 +360,206 @@ Validation has passed. Now our colleagues can feedback on our changes and if eve
...
+
+We can apply the changes by using the publish command without the `--dry-run` flag
+
+```bash
+snowplow-cli dp publish
+```
+
+### Add data products validation and publishing in the github actions
+
+Now that we've modeled a source app, data product and event spec, let's see how we can add them to the existing github actions workflows for data structures. You can customize your setup, use a separate repository or a separate actions, but in this example we'll add the data products publishing into the existing workflows.
```
Comment (cksnp) [inline docs/recipes/recipe-data-structures-in-git/index.md:506]:
```suggestion
Now that we've modeled a source application, data product and event specification, let's see how we can add them to the existing github actions workflows for data structures. You can customize your setup, use a separate repository or a separate actions, but in this example we'll add the data products publishing into the existing workflows.
```

File: docs/recipes/recipe-data-structures-in-git/index.md [hunk 12]
```diff
@@ -360,7 +360,206 @@ Validation has passed. Now our colleagues can feedback on our changes and if eve
...
+apiVersion: v1
+resourceType: data-product
+resourceName: 0edb4b95-3308-40c4-b266-eae2910d5d2a
+data:
+    name: Login
+    sourceApplications: []
+    eventSpecifications: []
+```
+
+Let's amend it to add an event spec, and a reference to a source app
```
Comment (cksnp) [inline docs/recipes/recipe-data-structures-in-git/index.md:458]:
```suggestion
Let's amend it to add an event specification, and a reference to a source application:
```

File: docs/recipes/recipe-data-structures-in-git/index.md [hunk 13]
```diff
@@ -360,7 +360,206 @@ Validation has passed. Now our colleagues can feedback on our changes and if eve
...
+      - uses: actions/checkout@v4
+
+      - uses: snowplow-product/setup-snowplow-cli@v1
+
+      - run: snowplow-cli ds validate --gh-annotate
+
+      - run: snowplow-cli dp publish --dry-run --gh-annotate
+```
+
+Data products, source apps and event specs don't have the dev and prod environments, so it's enough to publish them once.
```
Comment (cksnp) [inline docs/recipes/recipe-data-structures-in-git/index.md:533]:
```suggestion
Data products, source applications and event specifications don't have the dev and prod environments, so it's enough to publish them once.
```

File: docs/recipes/recipe-data-structures-in-git/index.md [hunk 14]
```diff
@@ -360,7 +360,206 @@ Validation has passed. Now our colleagues can feedback on our changes and if eve
...
+      SNOWPLOW_CONSOLE_API_KEY: ${{ secrets.SNOWPLOW_CONSOLE_API_KEY }}
+
+    steps:
+      - uses: actions/checkout@v4
+
+      - uses: snowplow-product/setup-snowplow-cli@v1
+
+      - run: snowplow-cli ds publish dev --managed-from $GITHUB_REPOSITORY
+
+      - run: snowplow-cli dp publish --gh-annotate
```
Comment (cksnp) [inline docs/recipes/recipe-data-structures-in-git/index.md:556]: Is `--gh-annotate` needed here?

File: docs/recipes/recipe-data-structures-in-git/index.md [hunk 15]
```diff
@@ -360,7 +360,206 @@ Validation has passed. Now our colleagues can feedback on our changes and if eve
...
+```
+
+You might want to publish data products in the `.github/workflows/publish-production.yml` as well, or only there. It depends on your setup, but if you strictly follow the rules and always merge to `main` from `develop`, the setup above should be enough.
+
 ## Let's break down what we've done
 
 * We have seen how snowplow-cli can be used to work with data structures from the command line
 * We have applied that knowledge to build github workflows which support automated validation and publication
+* We have added source apps, data products and event specs to use the same approach as data structures
```
Comment (cksnp) [inline docs/recipes/recipe-data-structures-in-git/index.md:565]:
```suggestion
* We have added source applications, data products and event specifications to use the same approach as data structures.
```

File: docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+### Creating data product
+
+```bash
+./snowplow-cli dp generate --data-product my-data-product
+
+```
+
+This command creates a minimal data product template in a new file `./data-products/my-data-product.yaml`.
+
+### Creating source app
```
Comment (cksnp) [inline docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md:34]:
```suggestion
### Creating source application
```

File: docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+
+```bash
+./snowplow-cli dp generate --source-app my-source-app
+
+```
+
+This command creates a minimal source app template in a new file `./data-products/source-apps/my-source-app.yaml`.
```
Comment (cksnp) [inline docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md:41]:
```suggestion
This command creates a minimal source application template in a new file `./data-products/source-apps/my-source-app.yaml`.
```

File: docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+
+### Creating event spec
```
Comment (cksnp) [inline docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md:43]:
```suggestion
### Creating event specification
```

File: docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+
+To create an event spec, you need to modify the existing data-product file and add an event spec object. Here's a minimal example:
```
Comment (cksnp) [inline docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md:45]:
```suggestion
To create an event specification, you need to modify the existing data-product file and add an event specification object. Here's a minimal example:
```

File: docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+resourceName: 3d3059c4-d29b-4979-a973-43f7070e1dd0
+data:
+    name: test-cli
+    sourceApplications:
+        - $ref: ./source-apps/my-source-app.yaml
+```
+
+### Modifying the event specifications source applications
+
+By default event specs inherit all the source apps of the data product. If you want to customise it, you can use the `excludedSourceApplications` in the event spec description to remove a given source application from an event spec
```
Comment (cksnp) [inline docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md:83]:
```suggestion
By default event specifications inherit all the source applications of the data product. If you want to customise it, you can use the `excludedSourceApplications` in the event specification description to remove a given source application from an event spec.
```

File: docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+          event:
+            source: iglu:com.snowplowanalytics.snowplow/button_click/jsonschema/1-0-0
+        - resourceName: b9c994a0-03b2-479c-b1cf-7d25c3adc572
+          name: Not quite everything
+          excludedSourceApplications:
+            - $ref: ./source-apps/specific.yaml
+          event:
+            source: iglu:com.snowplowanalytics.snowplow/button_click/jsonschema/1-0-0
+```
+In this example event spec `All source apps` is related to both `generic` and `specific` source apps, but event spec `Not quite everything` is related only to the `generic` source app
```
Comment (cksnp) [inline docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md:106]:
```suggestion
In this example event specification `All source apps` is related to both `generic` and `specific` source apps, but event specification `Not quite everything` is related only to the `generic` source application.
```

File: docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+
+
+### Downloading data products, event specifications and source apps
+
+```bash
+./snowplow-cli dp download
+```
+
+This command retrieves all organization data products, event specifications, and source apps. By default, it creates a folder named `data-products` in your current working directory. You can specify a different folder name as an argument if needed.
```
Comment (cksnp) [inline docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md:115]:
```suggestion
This command retrieves all organization data products, event specifications, and source applications. By default, it creates a folder named `data-products` in your current working directory. You can specify a different folder name as an argument if needed. 
```

File: docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+
+```bash
+./snowplow-cli dp download
+```
+
+This command retrieves all organization data products, event specifications, and source apps. By default, it creates a folder named `data-products` in your current working directory. You can specify a different folder name as an argument if needed. 
+
+The command creates the following structure:
+- A main `data-products` folder containing your data product files
+- A `source-apps` subfolder containing source app definitions
```
Comment (cksnp) [inline docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md:119]:
```suggestion
- A `source-apps` subfolder containing source application definitions
```

File: docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+- Event specifications embedded within their related data product files
```
Comment (cksnp) [inline docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md:120]:
```suggestion
- Event specifications embedded within their related data product files.
```

File: docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+
+
+### Validating data products, event specifications and source apps
```
Comment (cksnp) [inline docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md:123]:
```suggestion
### Validating data products, event specifications and source applications
```

File: docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+
+```bash
+./snowplow-cli dp validate
+```
+
+This command scans all files under `./data-products` and validates them using the BDP console. It checks:
+
+1. Whether each file is in a valid format (YAML/JSON) with correctly formatted fields
+2. Whether all source app references in the data product files are valid
```
Comment (cksnp) [inline docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md:132]:
```suggestion
2. Whether all source application references in the data product files are valid
```

File: docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md [hunk 27]
```diff
@@ -0,0 +1,145 @@
...
+3. Whether event specification rules are compatible with their schemas
+
+If validation fails, the command displays the errors in the console and exits with status code 1.
+
+
+### Publishing data products, event specifications and source apps
```
Comment (cksnp) [inline docs/understanding-tracking-design/defining-the-data-to-collect-with-data-poducts/cli/index.md:138]:
```suggestion
### Publishing data products, event specifications and source applications
```

**Review by cksnp — CHANGES_REQUESTED** [review]
Most of my comments are just about adding some punctuation or renaming apps to applications and specs to specifications. There's also one about a potential copy/paste oversight (not sure). Overall I think it's fine, so we can merge after those small things are addressed.

File: docs/recipes/recipe-data-structures-in-git/index.md (hunk 14 above)
Comment (gleb-lobov) [reply docs/recipes/recipe-data-structures-in-git/index.md:556]: Nope, noting to annotate on a merged PR, good catch

**Review by cksnp — APPROVED** [review]

**Comment by cksnp** [comment]
@gleb-lobov approved, do I merge as well?

**Pull request merged**